awx/api/views/__init__.py

```diff
@@ -3824,6 +3824,82 @@ def get_queryset(self):
         return job.get_event_queryset().select_related('host').order_by('start_line')
 
 
+class JobJobEventsChildrenSummary(APIView):
+    '''
+    Special view to facilitate processing job output in the UI.
+    In order to collapse events and their children, the UI needs to know how
+    many children exist for a given event.
+    This view returns a JSON object where the key is the event counter, and the value
+    is the number of children (and grandchildren) events.
+    Only events with children are included in the output.
+
+    e.g. Demo Job Template job
+    given:
+        tuple(event counter, uuid, parent_uuid)
+        (1, '4598d19e-93b4-4e33-a0ae-b387a7348964', '')
+        (2, 'aae0d189-e3cb-102a-9f00-000000000006', '4598d19e-93b4-4e33-a0ae-b387a7348964')
+        (3, 'aae0d189-e3cb-102a-9f00-00000000000c', 'aae0d189-e3cb-102a-9f00-000000000006')
+        (4, 'f4194f14-e406-4124-8519-0fdb08b18f4b', 'aae0d189-e3cb-102a-9f00-00000000000c')
+        (5, '39f7ad99-dbf3-41e0-93f8-9999db4004f2', 'aae0d189-e3cb-102a-9f00-00000000000c')
+        (6, 'aae0d189-e3cb-102a-9f00-000000000008', 'aae0d189-e3cb-102a-9f00-000000000006')
+        (7, '39a49992-5ca4-4b6c-b178-e56d0b0333da', 'aae0d189-e3cb-102a-9f00-000000000008')
+        (8, '504f3b28-3ea8-4f6f-bd82-60cf8e807cc0', 'aae0d189-e3cb-102a-9f00-000000000008')
+        (9, 'a242be54-ebe6-4021-afab-f2878bff2e9f', '4598d19e-93b4-4e33-a0ae-b387a7348964')
+
+    output:
+        {
+        "1": 8,
+        "2": 6,
+        "3": 2,
+        "6": 2
+        }
+    event counter 1 has 8 children
+    event counter 2 has 6 children
+    etc.
+    '''
+
+    def get(self, request, **kwargs):
+        job = get_object_or_404(models.Job, pk=kwargs['pk'])
+        if job.status in ("waiting", "pending", "running"):
+            return Response({})
+        events = job.get_event_queryset().values_list('counter', 'uuid', 'parent_uuid').order_by('counter')
+        # key is uuid, value is number of total children (including children of children, etc.)
+        map_uuid_children_tally = dict()
+        # key is uuid, value is event counter (the event order)
+        map_uuid_counter = dict()
+        # key is uuid, value is parent uuid. Used as a quick lookup
+        map_uuid_puuid = dict()
+        # ultimately we want to return dict(counter: tally), not dict(uuid: tally),
+        # so we build up this dict from the map_uuid_children_tally
+        map_counter_children_tally = dict()
+
+        def update_grandparents(uuid):
+            # recursive function that updates the tally for grandparent events
+            parent = map_uuid_puuid.get(uuid, None)
+            if parent is None:
+                return
+            map_uuid_children_tally[parent] += 1
+            update_grandparents(parent)
+
+        for counter, uuid, puuid in events:
+            if not uuid:
+                continue
+            map_uuid_counter[uuid] = counter
+            if not puuid:
+                continue
```

Comment:
For these cases, it would be better to filter out empty/null values when creating the original queryset. You're already at risk of performance problems, and that _could_ help.

Reply:
good call